content/blog/2022-05-19-A_Type_and_Effect_System_for_Deterministic_Parallel_Java/index.md

```diff
@@ -0,0 +1,123 @@
+
+```
+// take the "global" region P for this class as a template argument.
+class TreeNode<region P> {
+     region Links, L, R;  // create DPJ regions
+     // annotate that the member variable `mass` is in region P.
+     double mass in P:M ; // add P:M showing nesting of M under P
+     // annotate that member variables `left` and `right` are in region `Links`.
+     TreeNode<P:L> left in Links; // add P:L showing nesting of L under P
+     TreeNode<P:R> right in Links ; // add P:R showing nesting of R under P
+     // Annotate effect for the function `setMass` saying that it writes to region P.
+     void setMass(double mass) writes P { this.mass = mass; }
+     void initTree(double mass) {
+       cobegin {
+         // overal effect: reads Links writes L
+         left.mass = mass;
+         // overal effect: reads Links writes R
+         right.mass = mass;
+       }
+    }
+ }
+```
+
+The above code results in the following recursive region nesting as shown in the figure below. Here, the mass region M is different for every nested/recursive node.
+
+<p align="center">
+<img src="figure_1.jpg" alt="alt_text" title="image_tooltip" width="500" />
+</p>
+
+### Subtyping
+
+The DPJ type system supports subtyping to better express nested regions. A summary of the subtyping rules is shown below:
+
+* Two regions are nested within each other, R1 ≤ R2,  iff R1 is contained within R2, e.g. P:M ≤ P;
+* C<R1> is a subtype of C<R2> iff every region of R1 is in R2 (R1 ⊆ R2); examples:
+C<P:M> is a subtype of C<P:*>
+C<P:L:*:M> is a subtype of C<P:*:M>
+C<P:M> is not a subtype of C<P>
+* B<r> is a subtype of A<r> iff  B<region R> extends A<R> {} is sound, but B<r> is not a subtype of Object
+
+### Supporting Concurrency in Arrays
+
+In order to allow safe concurrency in arrays, the authors define two novel capabilities: (1) **index-parameterized arrays** and (2) **subarrays**.
+
+##### Index-Parameterized Arrays
+
+Index-Parameterized Arrays allow to describe concurrent operations on different elements of an array. The problem seems to be trivial for arrays of primitive types, however, it is way harder for arrays of **mutable references**. The idea behind index-parameterized arrays is simple: we need to parameterize each object reference in the array by the array index's integer expression. For example, the aforementioned ```TreeNode``` class should be parameterized as follows ```TreeNode<Root:[10]>``` for the element ```10``` in the array. The compiler then will check (with a symbolic integer expression solver) that all array accesses in the program going to the element parameterized by expression [i] happen through the array element ```i```. If this holds, we get formally proved safe concurrent access to all the distinct elements in the array, even with mutable references.
```

Comment:
TBH I don't quite get why the problem is trivial for arrays of primitives… isn't mutating the array *itself* a problem (like `a[100] = 5` for an `int[]` array)?

Reply:
I think, what they mean is that when we have an array of values (not references), and, say, multiple workers accessing these elements, determinism is guaranteed if these workers access different indexes, which is "easy" to check. I'm not exactly sure how it works, but I can imagine it's trivial for simple kernels, based on the loop bound conditions.

But for arrays of references, unique traversal is not enough, as the references (and nested references deeper in the recursion)  might change as well.